src/rest/actions/action.hpp

```diff
@@ -0,0 +1,280 @@
+     *
+     * @return  The timeout of the action.
+     */
+    Seconds GetTimeout(void) const { return mTimeout; }
+
+    /**
+     * Returns true if now is after the timeout of this action.
+     *
+     * @note This will not in any way check the status of the action. It is
+     *       purely comparing timepoints.
+     *
+     * @retval  True   If now is after the timeout of this action.
+     * @retval  False  If now is before or equal the timeout of this action.
+     */
+    bool IsBeyondTimeout(void) const { return mCreatedSteady + mTimeout < Clock::now(); }
+
+    /**
+     * Get the status of the action.
+     *
+     * @return  The status of the action.
+     */
+    ActionStatus GetStatus(void) const { return mStatus; }
+
+    /**
+     * Get the status of the action as a string.
+     *
+     * @return  The status of the action as a string.
+     */
+    const char *GetStatusString(void) const { return StatusToString(mStatus); }
+
+    /**
+     * Is the action pending or active?
+     *
+     * @retval True  If the action is pending or active.
+     * @retval False If the action is not pending or active.
+     */
+    bool IsPendingOrActive(void) const { return mStatus == kActionStatusPending || mStatus == kActionStatusActive; }
+
+    /**
+     * Convert the status of the action to string.
+     *
+     * @param[in] aStatus  The status of the action.
+     * @return  The status of the action as a string.
+     *
+     */
+    static const char *StatusToString(ActionStatus aStatus);
+
+protected:
+    static constexpr Seconds kDefaultTimeout{60};
+
+    /**
+     * Attempts to read the timeout property in a actions attribute cJSON
+     * object.
+     *
+     * If the timeout property cannot be found aTimeout remains unchanged.
+     *
+     * @param       aJson     The attributes cJSON object.
+     * @param[out]  aTimeout  Written with the timeout value if found.
+     *
+     * @retval  OT_ERROR_NONE       The timeout property was found and written to aTimeout
+     * @retval  OT_ERROR_PARSE      The timeout property was found but could not be parsed
+     * @retval  OT_ERROR_NOT_FOUND  The timeout property was not found or could not be parsed
+     */
+    static otError ReadTimeout(const cJSON *aJson, Seconds &aTimeout);
+
+    /**
+     * Attempts to read the timeout property in a actions attribute cJSON
+     * object.
+     *
+     * If the timeout property cannot be found or parsed aDefault is returned.
+     *
+     * @param       aJson         The attributes cJSON object.
+     * @param[in]   aDefault      The default value to return if the timeout property is not found.
+     *
+     * @return  The timeout value if found, otherwise aDefault.
+     */
+    static Seconds ReadTimeoutOrDefault(const cJSON &aJson, Seconds aDefault);
+
+    /**
+     * Attempts to read the destination property in a actions attribute cJSON
+     *
+     * @param       aJson     The attributes cJSON object.
```

Comment:
can this be null? If not, should we use reference instead of pointer?

Reply:
no, cannot be null. Changed to reference. Thx.